tactics/btermdn.ml

```diff
@@ -125,18 +155,77 @@ let constr_pat_discr env ts p =
       Some (GRLabel ref, stack)
     | PVar v when evaluable_named v env ts -> None
     | PVar v -> Some (GRLabel (VarRef v), stack)
-    | PProd (_,d,c) when stack = [] -> Some (ProdLabel, [d ; c])
+    | PProd (_,d,c) when stack = [] -> Some (ProdLabel, [Pattern d ; Pattern c])
+    | PLambda (_,d,c) when List.is_empty stack ->
+      begin
+        (* Check if the [t] could possibly unify with [fun x => f x].
+           Only return [LamLabel] if that is impossible! *)
+        match decomp_aux stack c with
+        | Some _ as res -> Some (LamLabel, Pattern d :: PartialPat res :: stack)
+        | None -> None
+      end
     | PSort s when stack = [] -> Some (SortLabel, [])
+    | PCase(_,_,p,_) | PIf(p,_,_) ->
+      begin
+        match decomp_aux stack p with
+        | Some (GRLabel (ConstructRef _), _) -> None (* over-approximating w.r.t. [fMATCH] *)
+        | Some _ as res -> Some (CaseLabel, PartialPat res :: stack)
+        | None -> None
+      end
     | _ -> None
+  and decomp (stack : partial_pat list) (t : partial_pat) : pat_res =
+    match t with
+    | Pattern p -> decomp_aux stack p
+    | PartialPat res -> res
   in
-  decomp [] (eta_reduce_pat p)
+  decomp [] p
+
+let constr_pat_discr_syntactic env p =
+  let open GlobRef in
+  let rec decomp_aux (stack : partial_pat list) (p : constr_pattern) : pat_res =
+    match eta_reduce_pat p with
+    | PApp (f,args) -> decomp_aux ((List.map (fun p -> Pattern p) @@ Array.to_list args) @ stack) f
+    | PProj (p,c) -> Some (ProjLabel p, Pattern c :: stack)
+    | PRef ((IndRef _) as ref)
+    | PRef ((ConstructRef _ ) as ref) ->
+      let ref = Environ.QGlobRef.canonize env ref in
+      Some (GRLabel ref, stack)
+    | PRef ((VarRef _) as ref) -> Some (GRLabel ref, stack)
+    | PRef ((ConstRef _) as ref) ->
+      let ref = Environ.QGlobRef.canonize env ref in
+      Some (GRLabel ref, stack)
+    | PVar v -> Some (GRLabel (VarRef v), stack)
+    | PProd (_,d,c) when stack = [] -> Some (ProdLabel, [Pattern d ; Pattern c])
+    | PLambda (_,d,c) when List.is_empty stack ->
+      begin
+        (* Check if the [t] could possibly unify with [fun x => f x].
+           Only return [LamLabel] if that is impossible! *)
+        match decomp_aux stack c with
```

Comment:
> I think this is useless. If a Hint Extern demands fun x => f x that's the only thing that matches. TC search might try to approximate eta but constr_pattern matching definitely doesn't.

Okay, so this is actually more complicated and I just remembered that I put that case there for a very good reason: eta normalization in `constr_val_discr` means that goals that contain `fun x => f x` actually end up looking for `f` instead. If `f` is opaque it has no chance of matching a `Hint Extern` registered on `fun x => f x`.

It's a weird situation. The goal is exactly `fun x => f x`, the hint is registered exactly (and syntactically) on `fun x => f x` but to make it all work out we need to make the hint match `f` behind the scenes (in practice that's probably going to be `_` in many cases) because everything is eta reduced. I wonder if there is a better way to do this. If there is, I don't think it will go in this MR so I'll resolve this part for now.